### ISSUE TITLE
Encrypting and decrypting a file

### ISSUE THREAD

**Issue by thisisrey** [voxpupuli/hiera-eyaml#244]
Tried encrypting a file , but only encrypts the content of it , not the file itself., what if the file consist of binary items like certificates file like .pfx?

**Comment by rnelson0** [comment]
Can you be more specific about what you're attempting, what commands you used and what the inputs and outputs look like? Please don't share private information but a similar, public file would help us replicate your issue.

**Comment by thisisrey** [comment]
I am encrypting a file via the command (eyaml encrypt -f filename), it turns out the only encrypted is the content, how should we encrypt the whole file?, or if ever is there a way where in you can encrypt .pfx(certificte file which consist of binary) and then decrypt it through puppet manifest code, like for example calling the encrypted values and passing it to the puppet file resource?

**Comment by Cinderhaze** [comment]
You can have the (encrypted) file contents in hiera, and use it as the 'content' param on your file resource

**Comment by jhg03a** [comment]
Maybe a bigger question, does hiera-eyaml support encrypt/decrypt operations on non-text files?  I know it works fine on things like PEM formatted certificates and key files with the -f argument.  To check you could get a MD5sum of the binary file, run it through eyaml encrypt & decrypt, and then check the MD5sum on the other side.  I will say you end up with some fairly nasty encrypted strings in hiera if you do this for a file of any significant size.  In my use case, I put my wildcard SSL certificates through eyaml and set them in my common yaml file at the bottom of my hierarchy so I can reuse it in many places.  I will also give the warning that you should be careful encrypting some kinds of files without the -f argument because changing the line endings can cause problems with some types of files like pem formatted keys.

**Comment by rnelson0** [comment]
@thisisrey I'm not sure what you mean when you say only the content is encrypted, not the entire file. What exactly are you seeing and what are you expecting?

I do believe that @Cinderhaze's suggestion should suffice for you.

**Comment by rnelson0** [comment]
@jhg03a I don't believe there's explicit binary support, as you mentioned there's a lot of caveats there and encryption in general often results in lossy decryption later due to those issues.

**Comment by jhg03a** [comment]
It's important to separate eyaml from hiera-eyaml in this case.  If eyaml can serialize and deserialize a binary file properly using the -f argument and validating with md5 hashes, then hiera-eyaml should be able to do the same by adding the serialized version of the binary file as a string and using the "content" argument of a puppet file resource combined with a hiera lookup.

**Comment by jhg03a** [comment]
It's not a great solution and I wouldn't consider it a best practice, but it's a workable solution as far as I'm concerned.

**Comment by jhg03a** [comment]
This also will not preserve any kind of file properties of the original file, you'll have to model those separately in your puppet code.  For example, you'll lose any kind of Posix ACL.

**Comment by lukastopiarz** [comment]
Any plans to implement this feature? It would be very helpfull.
